src/compute/src/render/reduce.rs

```diff
@@ -317,92 +319,138 @@ where
-                        input.len(), distinct_aggregate_types.len());
-                    soft_assert_or_log!(false, "Mismatched aggregates for key in ReduceCollation");
                 }
-            }
-        })
+            },
+            move |key, input, output| {
+                if input.len() != n_distinct_aggregate_types {
+                    // We expected to stitch together exactly as many aggregate types as requested
+                    // by the collation. If we cannot, we log an error and produce no output for
+                    // this key.
+                    let message = "Mismatched aggregates for key in ReduceCollation";
+                    warn!(
+                        ?key,
+                        debug_name,
+                        n_aggregates_requested = input.len(),
+                        n_distinct_aggregate_types,
+                        "[customer-data] {message}"
+                    );
+
+                    error!("{message}");
+                    output.push((EvalError::Internal(message.to_string()).into(), 1));
+                    return;
+                }
+
+                let mut accumulable = DatumList::empty().iter();
+                let mut hierarchical = DatumList::empty().iter();
+                let mut basic = DatumList::empty().iter();
+                for ((reduction_type, row), _) in input.iter() {
+                    match reduction_type {
+                        ReductionType::Accumulable => accumulable = row.iter(),
+                        ReductionType::Hierarchical => hierarchical = row.iter(),
+                        ReductionType::Basic => basic = row.iter(),
+                    }
+                }
+
+                for typ in aggregate_types_err.iter() {
+                    let datum = match typ {
+                        ReductionType::Accumulable => accumulable.next(),
+                        ReductionType::Hierarchical => hierarchical.next(),
+                        ReductionType::Basic => basic.next(),
+                    };
+                    if datum.is_some() {
+                        continue;
+                    }
+
+                    // We cannot properly reconstruct a row if aggregates are missing.
+                    // This situation is not expected, so we log an error if it occurs.
+                    let message = "Missing value for key in ReduceCollation";
+                    warn!(?typ, ?key, debug_name, "[customer-data] {message}");
+                    error!("{message}");
+                    output.push((EvalError::Internal(message.to_string()).into(), 1));
+                    return;
+                }
+
+                // Note that we also do not want to have anything left over to stich.
+                // If we do, then we also have an error and would violate join semantics.
+                if accumulable.next() == None && hierarchical.next() == None && basic.next() == None
```

Comment:
No strong opinion, but could do here and above:
```suggestion
                if (accumulable.next(), hierarchical.next(), basic.next()) == (None, None, None)
```